nixos/systemd-boot: Fix installed version regexp

jrobsonchase opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
The regexp was only matching numbers and not the '.', so everyone using
systemd-boot would always see `could not find any previously installed
systemd-boot` on a `nixos-rebuild`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  ~- [ ] x86_64-darwin~
  ~- [ ] aarch64-darwin~
~- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))~
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"` *<- Not sure if this is relevant?*
~- [ ] Tested execution of all binary files (usually in `./result/bin/`)~
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  ~- [ ] (Package updates) Added a release notes entry if the change is major or breaking~
  ~- [ ] (Module updates) Added a release notes entry if the change is significant~
  ~- [ ] (Module addition) Added a release notes entry if adding a new NixOS module~
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jrobsonchase committed:
nixos/systemd-boot: Fix installed version regexp

The regexp was only matching numbers and not the '.', so everyone using
systemd-boot would always see `could not find any previously installed
systemd-boot` on a `nixos-rebuild`.

jrobsonchase committed:
nixos/systemd-boot: Update test for version regexp

This should fail since the regexp in systemd-boot-builder.py won't match
the '.' in the version string.

SuperSandro2000 commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -218,7 +218,7 @@ def main() -> None:
 
         # See status_binaries() in systemd bootctl.c for code which generates this
-        m = re.search("^\W+File:.*/EFI/(BOOT|systemd)/.*\.efi \(systemd-boot (\d+)\)$",
+        m = re.search("^\W+File:.*/EFI/(BOOT|systemd)/.*\.efi \(systemd-boot ([0-9.]+)\)$",
```
```suggestion
        m = re.search("^\W+File:.*/EFI/(BOOT|systemd)/.*\.efi \(systemd-boot (\d\.\d+)\)$",
```

pennae replied:
sounds like a bad idea, our system has:
```
File: └─/EFI/systemd/systemd-bootx64.efi (systemd-boot 247.4-2-arch)
```

it's a lustrated system (from arch), maybe matching `systemd-boot (\d+\.\d+)[^)]*\)$` is better

jrobsonchase replied:
@SuperSandro2000 That's not going to work - There can be multiple digits before the `.`, and I don't know that systemd's versioning scheme guarantees exactly one `.`. Seems reasonable to allow for more version levels, especially since they're just being compared with string ordering.

@pennae Good point, I had a similar one installed earlier today. I'll add the `[^)]*` :+1:

jrobsonchase commented:
Changed the `0-9` to `\d` and added `[^)]+` outside of the capture group. Ready for another review pass.

I'm still on the fence about capturing vs not capturing the extra suffix like in the arch case. Could be swayed either way.

I'm also not sure that [this check](https://github.com/NixOS/nixpkgs/blob/b8ad1c3282774603042b3c7bc2aed74bb5fe1cdb/nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py#L227) is actually correct. Should it be "always upgrade and never downgrade" or "make it match the version the module expects?" If the later, an equivalence check would be better than the string order.

sugar700 reviewed (changes requested):
That doesn't seem right. Later in this code the result of this check is used for string comparison. For instance, if a current version of systemd is 249.9, and we are going to install 249.10 then the comparison will fail.

```python
>>> "249.10" > "249.9"
False
```

(to be fair, the code below was wrong previously, however it likely wouldn't cause issues until release of systemd 1000)

jrobsonchase commented:
Yeah, that's true. We'd need to parse the version as either a float if we assume it'll only ever contain one `.`, or a semver parser otherwise.

> wouldn't cause issues until release of systemd 1000

That assumes that it ever actually got to that check, which it wouldn't :stuck_out_tongue: 

Is there any reason we shouldn't just capture the entire version string and ensure that it always completely matches the version from nixpkgs? It might even be prudent to add a hash to the end of our version to make sure that it's actually the nixos binary and not just one that shares the same version.

danielfullmer commented:
Just to add a bit of context here since I originally wrote this code:  I did incorrectly assume the version string would always just be a series of digits, however, if it was originally installed by NixOS then it would be.  From a closer look at the code (`boot/efi/boot.c` line 28), we see that it is set to whatever `GIT_VERSION` is set to:
```c
static const char _used_ _section_(".sdmagic") magic[] = "#### LoaderInfo: systemd-boot " GIT_VERSION " ####";
```

`GIT_VERSION` is set via `src/version/version.h.in` from `@VCS_TAG@`, which can be set via meson using `-Dversion-tag`.  (This is how arch sets their custom version tag).

Another thing is that `bootctl update` itself will only update the boot loader if it detects that its version is newer than the one present.  (see `src/boot/bootctl.c`, and track how `verb_install` will indirectly call  `copy_file_with_version_check`, which calls down to `compare_version`. Earlier this year it looks like they replaced the simple version comparator with a more complicated one (`strversmp_improved`).

Since `bootctl update` will already do the version comparison for us, and duplicating that functionality in our code would be error prone, perhaps we can just always run `bootctl update`?  I don't recall my original reasoning for duplicating the version comparison in `systemd-boot-builder.py`, but if it's not needed then we should try to get rid of it.

jrobsonchase commented:
> Since bootctl update will already do the version comparison for us, and duplicating that functionality in our code would be error prone, perhaps we can just always run bootctl update? I don't recall my original reasoning for duplicating the version comparison in systemd-boot-builder.py, but if it's not needed then we should try to get rid of it

That sounds like the right approach to me. At worst, the bootctl version check will get it wrong and install an older version from nixpkgs, which is arguably more correct anyway.

jrobsonchase committed:
nixos/systemd-boot: Remove the installed version check altogether

bootctl does it as a part of its update process anyway, so we're just
duplicating code.

danielfullmer commented:
Looks like this commit would solve that issue, but I don't believe it's in the current systemd used in NixOS yet:
https://github.com/systemd/systemd/commit/e5a8b4b593e7da1823f3258f7b4e7f2d370a7c3c
And we should probably also add `--graceful` to the options.

colemickens commented:
This is interesting, this change has bigger ramifications:

>Ignore failure when the EFI System Partition cannot be found, when EFI variables cannot be written, or a different or newer boot loader is already installed. Currently only applies to random seed and update operations.

I think this would actually fix a case of NixOS installs failing on systems like... say the raspberry pi4, with u-boot ->efi-> `systemd-boot`. Currently that's not an option because without `--graceful`, the install fails because it can't access UEFI variables.

It seems like `graceful` might be ... too graceful?

I really don't get why `systemd` doesn't default to:
* warn and skip if the installed version is the same or older
* have a `--reinstall` flag to force reinstall
* leave `--graceful` the way it is

Seems like it would've allowed more appropriate, granular behavior.

jrobsonchase committed:
nixos/systemd-boot: Re-add the verison check, but as equivalence-only

jrobsonchase commented:
> Instead of comparing if the new version is greater, I think it would then be sufficient to just check if the version strings are different

This is exactly what I suggested [here](https://github.com/NixOS/nixpkgs/pull/140046#issuecomment-931444779), though I was unaware that `bootctl` had a version check to prevent downgrades. I'm fine with either approach ¯\\\_(ツ)_/¯

jrobsonchase committed:
nixos/systemd-boot: Use the correct version string from the bootctl --version output

jrobsonchase commented:
@SuperSandro2000 Is there anything left blocking this from being merged?

anund commented:
The change works :+1:. Working systemd-boot post ~248 exposes some users to https://github.com/systemd/systemd/issues/19191 (see #143847) :-1:. The documented way to rollback the bootloader is also broken #144007 :-1:

xaverdh commented:
> The change works +1. Working systemd-boot post ~248 exposes some users to [systemd/systemd#19191](https://github.com/systemd/systemd/issues/19191) (see #143847). The? documented way to rollback the bootloader is also broken #144007.

Should be patch our systemd with https://github.com/systemd/systemd/commit/e98d271e57f3d0356e444b6ea2d48836ee2769b0 before merging this?

xaverdh commented:
cc @flokli (re patching systemd)

flokli commented:
> cc @flokli (re patching systemd)

I replied on the issue for that (https://github.com/NixOS/nixpkgs/issues/143847#issuecomment-956006957). Let's do the discussion about patching or not there.

flokli commented:
I just merged https://github.com/NixOS/nixpkgs/pull/144422 into staging.

Could this be rebased on top of staging as well? I'd like to avoid this making `bootctl install` work again, and potentially breaking machines affected by https://github.com/NixOS/nixpkgs/issues/143847 until the fix from staging ends up in master.

jrobsonchase commented:
> Could this be rebased on top of (and target) staging as well?

Done!

flokli commented:
Thanks!

flokli commented:
Those should now both end in the staging-next cycle after https://github.com/NixOS/nixpkgs/pull/144730.

Profpatsch commented:
WHY IS THIS PARSED ANYWAY

I’m sorry but why the heck are we parsing display output of bootctl? Is there no machine-readable output format?
